iceoryx_posh/include/iceoryx_posh/popo/trigger_handle.hpp

```diff
@@ -0,0 +1,82 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_POSH_POPO_TRIGGER_HANDLE_HPP
+#define IOX_POSH_POPO_TRIGGER_HANDLE_HPP
+
+#include "iceoryx_posh/internal/popo/building_blocks/condition_variable_data.hpp"
+#include "iceoryx_posh/popo/trigger.hpp"
+#include "iceoryx_utils/cxx/method_callback.hpp"
+
+#include <mutex>
+
+namespace iox
+{
+namespace popo
+{
+/// @brief TriggerHandle is threadsafe without restrictions in a single process.
+///        Not qualified for inter process usage.
+class TriggerHandle
+{
+  public:
+    TriggerHandle() = default;
+    /// @brief Creates a TriggerHandle
+    /// @param[in] conditionVariableDataPtr pointer to a condition variable data struct
+    /// @param[in] resetCallback callback which will be called it goes out of scope or reset is called
+    /// @param[in] uniqueTriggerId the unique trigger id of the Trigger which corresponds to the TriggerHandle. Usually
+    /// stored in a Notifyable. It is required for the resetCallback
+    TriggerHandle(ConditionVariableData* const conditionVariableDataPtr,
+                  const cxx::MethodCallback<void, uint64_t> resetCallback,
+                  const uint64_t uniqueTriggerId) noexcept;
+    TriggerHandle(const TriggerHandle&) = delete;
+    TriggerHandle& operator=(const TriggerHandle&) = delete;
+
+    TriggerHandle(TriggerHandle&& rhs) noexcept;
+    TriggerHandle& operator=(TriggerHandle&& rhs) noexcept;
+    ~TriggerHandle();
+
+    /// @brief returns true if the TriggerHandle is valid otherwise false. A TriggerHandle is valid if
+    /// m_conditionVariableDataPtr != nullptr.
+    explicit operator bool() const noexcept;
+
+    /// @brief returns true if the TriggerHandle is valid otherwise false. A TriggerHandle is valid if
+    /// m_conditionVariableDataPtr != nullptr.
+    bool isValid() const noexcept;
+
+    /// @brief triggers the Trigger and informs the Notifyable which verifies that the Trigger was triggered by calling
+    /// the hasTriggeredCallback
+    void trigger() noexcept;
+
+    /// @brief calls the resetCallback and invalidates the TriggerHandle
+    void reset() noexcept;
+
+    /// @brief invalidates the TriggerHandle without calling the reset callback
+    void invalidate() noexcept;
+
+    /// @brief returns the uniqueTriggerId
+    uint64_t getUniqueId() const noexcept;
+
+    /// @brief returns the pointer to the ConditionVariableData
+    ConditionVariableData* getConditionVariableData() noexcept;
+
+  private:
+    ConditionVariableData* m_conditionVariableDataPtr = nullptr;
```

Comment:
It needs to know the ConditionVariableData since we require this information for instance in the subscriber. Here we acquire this pointer and hand it out to the chunk sender which is using it to notify the waitset. But it would be great if we could remove this but this would be an optimization for the future.

At the moment the trigger is stored in a vector inside of the waitset and therefore a pointer to a trigger is not possible. The uniqueId  is somehow a replacement for that. I added this as an open point in the waitset issue: https://github.com/eclipse/iceoryx/issues/341 Maybe we switch to a list inside of the waitset and then we could get rid of this unique id again.